module/os/linux/zfs/zpl_inode.c

```diff
@@ -317,17 +317,19 @@ zpl_tmpfile(struct inode *dir, struct dentry *dentry, umode_t mode)
 		/* d_tmpfile will do drop_nlink, so we should set it first */
 		set_nlink(ip, 1);
 #ifndef HAVE_TMPFILE_DENTRY
-		d_tmpfile(file, ip);
-
 		error = zpl_xattr_security_init(ip, dir,
 		    &file->f_path.dentry->d_name);
+		if (error == 0)
+			error = zpl_init_acl(ip, dir);
+		if (error == 0)
+			d_tmpfile(file, ip);
 #else
-		d_tmpfile(dentry, ip);
-
 		error = zpl_xattr_security_init(ip, dir, &dentry->d_name);
-#endif
 		if (error == 0)
 			error = zpl_init_acl(ip, dir);
+		if (error == 0)
+			d_tmpfile(dentry, ip);
+#endif
 #ifndef HAVE_TMPFILE_DENTRY
```

Comment:
Let's simply this a bit more and pull  the `finish_open_simple` in the existing `#ifndef HAVE_TMPFILE_DENTRY` case above.